### PR TITLE
[CONTINT-4479] add crio workloadmeta collector

### PR DESCRIPTION
**Pull request by ewoodthomas**
### What does this PR do?

This PR adds a dedicated Workloadmeta collector for the CRI-O runtime. Prior to this, the only way to get container-level information for CRI-O runtimes was through the Kubelet.

### Motivation

https://datadoghq.atlassian.net/browse/CONTINT-4479

Expand support for commonly used container runtimes. The ultimate goal of adding a collector for CRI-O is to enable image collection for CRI-O runtimes.

### Describe how to test/QA your changes

**Test with cri-api v1alpha2 endpoint**

- Start a minikube cluster with a CRI-O runtime
```
minikube start --container-runtime=crio
```
- Verify your CRI-O version is v1.24 (should default to this).
```
kubectl get nodes -o wide
```
- Deploy the agent
- Check `workload-list` to verify that the collector has container entities coming from the CRI-O collector.
```
kubectl exec -it <agent_pod_name> -n <agent_pod_namespace> -- agent workload-list -v
```
- Search for `container source:runtime`. The entity should contain fields similar to this: 
- Note: earlier versions of CRI-O do not populate Status.Resources. Some info is parsed from Status.Info.
```
=== Entity container source:runtime id: bac37111e70cfa34a4ace958abed1d7ddd48ac943992d418c7b722c6d3a36b63 ===
----------- Entity ID -----------
Kind: container ID: bac37111e70cfa34a4ace958abed1d7ddd48ac943992d418c7b722c6d3a36b63
----------- Entity Meta -----------
Name: coredns
Namespace: kube-system
Annotations: io.kubernetes.container.terminationMessagePath:/dev/termination-log io.kubernetes.container.terminationMessagePolicy:File io.kubernetes.pod.terminationGracePeriod:30 io.kubernetes.container.ports:[{"containerPort":53,"name":"dns","protocol":"UDP"},{"containerPort":53,"name":"dns-tcp","protocol":"TCP"},{"containerPort":9153,"name":"metrics","protocol":"TCP"}] io.kubernetes.container.hash:e6f52134 io.kubernetes.container.restartCount:0
Labels: io.kubernetes.pod.uid:71307c95-fd39-47df-bd06-75e75372ddc2 io.kubernetes.container.name:coredns io.kubernetes.pod.name:coredns-5cfdc65f69-vhmk9 io.kubernetes.pod.namespace:kube-system
----------- Image -----------
Name: registry.k8s.io/coredns/coredns
Tag: v1.11.1
ID: 2437cf762177702dec2dfe99a09c37427a15af6d9a57c456b65352667c223d93
Raw Name: registry.k8s.io/coredns/coredns:v1.11.1
Short Name: coredns
Repo Digest: registry.k8s.io/coredns/coredns@sha256:1eeb4c7316bacb1d4c8ead65571cd92dd21e27359f0d4917f1a5822a73b75db1
----------- Container Info -----------
Runtime: cri-o
RuntimeFlavor:
Running: true
Status: running
Health:
Created At: 2024-11-01 15:59:54.898622 +0000 UTC
Started At: 2024-11-01 15:59:54.902917928 +0000 UTC
Finished At: 1970-01-01 00:00:00 +0000 UTC
Exit Code: 0
----------- Resources -----------
Hostname:
Network IPs:
PID: 0
Cgroup path:
----------- Ports -----------
Port: 53
Name: dns
Protocol: UDP
Host Port: 0
Port: 53
Name: dns-tcp
Protocol: TCP
Host Port: 0
Port: 9153
Name: metrics
Protocol: TCP
Host Port: 0
===
```
- `Ports` should only be populated if the ports are listed in the annotations.

**Test with cri-api v1 endpoint**
- Minikube has limited version control, so I used this hacky Kind setup to use a recent version of cri-o that uses the V1 endpoint of the cri-api
- Clone Kubernetes
```
mkdir -p /tmp/kind
cd /tmp/kind
git clone -b v1.30.0 git@github.com:kubernetes/kubernetes.git
kind build node-image /tmp/kind/kubernetes
```
- Create a Dockerfile to set CRI-O as the runtime
```
FROM kindest/node:latest

ARG CRIO_VERSION
ARG PROJECT_PATH=prerelease:/$CRIO_VERSION

RUN echo "Installing Packages ..." \
    && apt-get clean \
    && apt-get update -y \
    && DEBIAN_FRONTEND=noninteractive apt-get install -y \
    software-properties-common vim gnupg \
    && echo "Installing cri-o ..." \
    && curl -fsSL https://pkgs.k8s.io/addons:/cri-o:/$PROJECT_PATH/deb/Release.key | gpg --dearmor -o /etc/apt/keyrings/cri-o-apt-keyring.gpg \
    && echo "deb [signed-by=/etc/apt/keyrings/cri-o-apt-keyring.gpg] https://pkgs.k8s.io/addons:/cri-o:/$PROJECT_PATH/deb/ /" | tee /etc/apt/sources.list.d/cri-o.list \
    && apt-get update \
    && DEBIAN_FRONTEND=noninteractive apt-get --option=Dpkg::Options::=--force-confdef install -y cri-o \
    && echo "Configuring CRI-O runtime" \
    && sed -i 's/containerd/crio/g' /etc/crictl.yaml \
    && systemctl disable containerd \
    && systemctl enable crio
```
- Create a Kind node image from this dockerfile
```
CRIO_VERSION=v1.30
docker build --build-arg CRIO_VERSION=$CRIO_VERSION -t kindnode/crio:$CRIO_VERSION .
```
- Create a custom config `kind-crio.yaml` to use the cri-o runtime
```
kind: Cluster
apiVersion: kind.x-k8s.io/v1alpha4
nodes:
  - role: control-plane
    kubeadmConfigPatches:
      - |
        kind: InitConfiguration
        nodeRegistration:
          criSocket: unix:///var/run/crio/crio.sock
      - |
        kind: JoinConfiguration
        nodeRegistration:
          criSocket: unix:///var/run/crio/crio.sock
  - role: worker
    kubeadmConfigPatches:
      - |
        kind: JoinConfiguration
        nodeRegistration:
          criSocket: unix:///var/run/crio/crio.sock
```
- Create the Kind cluster
```
kind create cluster --image kindnode/crio:$CRIO_VERSION --config kind-crio.yaml
```
- Verify your CRI-O version is v1.30.
```
kubectl get nodes -o wide
```
- Deploy the agent
- Check `workload-list` to verify that the collector has container entities coming from the CRI-O collector.
```
kubectl exec -it <agent_pod_name> -n <agent_pod_namespace> -- agent workload-list -v
```
- Search for `container source:runtime`. The entity should contain fields similar to this: 
```
=== Entity container source:runtime id: bac37111e70cfa34a4ace958abed1d7ddd48ac943992d418c7b722c6d3a36b63 ===
----------- Entity ID -----------
Kind: container ID: bac37111e70cfa34a4ace958abed1d7ddd48ac943992d418c7b722c6d3a36b63
----------- Entity Meta -----------
Name: coredns
Namespace: kube-system
Annotations: io.kubernetes.container.terminationMessagePath:/dev/termination-log io.kubernetes.container.terminationMessagePolicy:File io.kubernetes.pod.terminationGracePeriod:30 io.kubernetes.container.ports:[{"containerPort":53,"name":"dns","protocol":"UDP"},{"containerPort":53,"name":"dns-tcp","protocol":"TCP"},{"containerPort":9153,"name":"metrics","protocol":"TCP"}] io.kubernetes.container.hash:e6f52134 io.kubernetes.container.restartCount:0
Labels: io.kubernetes.pod.uid:71307c95-fd39-47df-bd06-75e75372ddc2 io.kubernetes.container.name:coredns io.kubernetes.pod.name:coredns-5cfdc65f69-vhmk9 io.kubernetes.pod.namespace:kube-system
----------- Image -----------
Name: registry.k8s.io/coredns/coredns
Tag: v1.11.1
ID: 2437cf762177702dec2dfe99a09c37427a15af6d9a57c456b65352667c223d93
Raw Name: registry.k8s.io/coredns/coredns:v1.11.1
Short Name: coredns
Repo Digest: registry.k8s.io/coredns/coredns@sha256:1eeb4c7316bacb1d4c8ead65571cd92dd21e27359f0d4917f1a5822a73b75db1
----------- Container Info -----------
Runtime: cri-o
RuntimeFlavor:
Running: true
Status: running
Health:
Created At: 2024-11-01 15:59:54.898622 +0000 UTC
Started At: 2024-11-01 15:59:54.902917928 +0000 UTC
Finished At: 1970-01-01 00:00:00 +0000 UTC
Exit Code: 0
----------- Resources -----------
TargetMemoryLimit: 178257920
Hostname:
Network IPs:
PID: 0
Cgroup path:
----------- Ports -----------
Port: 53
Name: dns
Protocol: UDP
Host Port: 0
Port: 53
Name: dns-tcp
Protocol: TCP
Host Port: 0
Port: 9153
Name: metrics
Protocol: TCP
Host Port: 0
===
```
- `Ports` should only be populated if the ports are listed in the annotations.

### Additional Notes

**Event Stream**
Ideally, we would like to use the event stream for the cri-o collector, but there are a few caveats with it. The event stream (`GetContainerEvents`), is only supported by the cri-api `v1` endpoint and is currently an alpha feature. In order to use the event stream, we would need to first determine which endpoint the cri-api is using. In addition to this, `enable_pod_events` must be set to `true` (`false` by default) in the cri-o runtime config. Since this would probably be rarely setup and supporting the event stream adds a decent amount of complexity, I opted to not use it. It can be added in the future if necessary.

**Misc bits that may need to be changed**
1. The way the ports are added the workloadmeta container is a little hacky. The cri-api doesn't expose the container's ports, but I noticed Kubernetes adds annotations containing the ports to some of the containers. I scrape the annotations to get the list of ports. I left it in, but can remove it. I'll leave a comment in the code as well.
2. I added a `crio` build flag, but can switch it to the existing `cri` build flag if necessary. Not sure which option is best.

### CODE REVIEW

File: comp/core/workloadmeta/collectors/internal/crio/crio.go
Comment (ewoodthomas) [inline comp/core/workloadmeta/collectors/internal/crio/crio.go:279]: This way of getting container ports feels pretty hacky. The CRI-API does not expose container ports, but Kubernetes adds ports into the annotations for some of its pods like `coredns` and `kube-apiserver`. I added this in, but maybe it should just be removed.

File: releasenotes/notes/crio-workloadmeta-collector-52aa605d1f316532.yaml
Comment (drichards-87) [inline releasenotes/notes/crio-workloadmeta-collector-52aa605d1f316532.yaml:11]: Should "Workloadmeta" be two separate words?

**Review by drichards-87 — APPROVED** [review]
Left a question from Docs and approved the PR.

**Review by CelianR — APPROVED** [review]

**Review by KevinFairise2 — APPROVED** [review]

File: pkg/util/crio/crio_util.go
Comment (misteriaud) [inline pkg/util/crio/crio_util.go:1]: Since this util package is used only for containerized env, you should update `.github/CODEOWNERS` to add this rule:
```
/pkg/util/crio/                      @DataDog/container-integrations
```

Comment (ewoodthomas) [reply pkg/util/crio/crio_util.go:1]: Updated codeowners, but still seems to mark pkg/util/crio/crio_util.go as requiring agent-shared-components review

**Review by jeremy-hanna — APPROVED** [review]

File: pkg/util/crio/crio_util.go
Comment (jeremy-hanna) [reply pkg/util/crio/crio_util.go:1]: approved for ASC

**Review by clamoriniere — COMMENTED** [review]
small comments

File: releasenotes/notes/crio-workloadmeta-collector-52aa605d1f316532.yaml
Comment (ewoodthomas) [reply releasenotes/notes/crio-workloadmeta-collector-52aa605d1f316532.yaml:11]: Double check on past release notes, it should just be one word.

File: comp/core/workloadmeta/collectors/internal/crio/containers.go
Comment (sblumenthal) [inline comp/core/workloadmeta/collectors/internal/crio/containers.go:168]: nit: minor optimization, but if `imageSpec` is `nil` then I would expect nothing would be returned, so we could probably short circuit this call in that case to prevent unnecessary queries to the crio socket

File: comp/core/workloadmeta/collectors/internal/crio/containers.go
Comment (sblumenthal) [inline comp/core/workloadmeta/collectors/internal/crio/containers.go:37]: [Please update the comment](https://github.com/DataDog/datadog-agent/blob/86dab7cf210a986cb2e5d8ea6497476444f4c6ed/comp/core/workloadmeta/def/types.go#L61-L64) for `SourceRuntime` to include `crio`

File: pkg/util/crio/crio_util.go
Comment (sblumenthal) [inline pkg/util/crio/crio_util.go:33]: nit: The variable definitions separated out in this manner are not needed

**Review by sblumenthal — APPROVED** [review]
A few comments, but overall the structure looks good! Approving, but please fix the comments before merging

**Review by clamoriniere — APPROVED** [review]
/lgtm

Thanks for taking into consideration my comments 🙇

**Pull request merged**
